pkg/controller/common/encryptimage.go

```diff
@@ -0,0 +1,250 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/errors"
+	alicloudclient "github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud/client"
+	"github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud/client/ros"
+	gcorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/utils/retry"
+
+	_ "embed"
+)
+
+// CopyImageROSTemplate contains the content of CopyImage ROS template. https://www.alibabacloud.com/help/doc-detail/116189.htm?spm=a2c63.l28256.b99.201.713413a3FkLSIx
+//go:embed copyimage_ros.yaml
+var CopyImageROSTemplate string
+
+// UseEncryptedSystemDisk checks whether the given volume needs encryption
+// If volume is unexpected, it will returns error to avoid unexpected result
+func UseEncryptedSystemDisk(volume interface{}) (bool, error) {
+	if volume == nil {
+		return false, nil
+	}
+
+	if v, ok := volume.(*extensionsv1alpha1.Volume); ok {
+		return v.Encrypted != nil && *v.Encrypted, nil
+	}
+
+	if v, ok := volume.(*gcorev1beta1.Volume); ok {
+		return v.Encrypted != nil && *v.Encrypted, nil
+	}
+
+	return false, fmt.Errorf("type of input volume [%v] is unexpected", volume)
+}
+
+// ImageEncrypter declares interfaces to operate an encrypted image
+type ImageEncrypter interface {
+	TryToGetEncryptedImageID(ctx context.Context, timeout time.Duration, interval time.Duration) (string, error)
+}
+
+type imageEncryptor struct {
+	regionID      string
+	sourceImageID string
+	imageName     string
+	imageVersion  string
+	rosClient     alicloudclient.ROS
+}
+
+// NewImageEncryptor creates an ImageEncrypter instance
+func NewImageEncryptor(client alicloudclient.ROS, regionID, imageName, imageVersion, sourceImageID string) ImageEncrypter {
+	return &imageEncryptor{
+		regionID:      regionID,
+		sourceImageID: sourceImageID,
+		imageName:     imageName,
+		imageVersion:  imageVersion,
+		rosClient:     client,
+	}
+}
+
+// TryToGetEncryptedImageID will get image id from stack.
+// If the stack doesn't exit, it will create one and wait until the stack gets complete
+// It always take around 10 minutes to copy an encrypted image.
+// @Param timout is how long it will wait for stack to complete
+// @Param interval is time period to check whether the stack is ready via REST API
```

Comment:
```suggestion
// TryToGetEncryptedImageID will get image id from stack.
// If the stack doesn't exist, it will create one and wait until the stack creation is complete
// It always takes around 10 minutes to copy an encrypted image.
// @Param timeout is the maximum time for it to wait for stack creation to complete
// @Param interval is the time period to check whether the stack is ready via REST API
```